polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
-      values.push_back(ub_ops[idx + looplike.getUpperBoundsMap().getNumDims()]);
-    }
-
-    auto iset = IntegerSet::get(
-        /*dim*/ looplike.getLowerBoundsMap().getNumDims() +
-            looplike.getUpperBoundsMap().getNumDims(),
-        /*symbols*/ looplike.getLowerBoundsMap().getNumSymbols() +
-            looplike.getUpperBoundsMap().getNumSymbols(),
-        exprs, eqflags);
-    auto ifOp = b.create<AffineIfOp>(looplike.getLoc(), TypeRange(), iset,
-                                     values, /*else*/ false);
-    looplike->moveBefore(ifOp.getThenBlock()->getTerminator());
   }
-  for (auto op : opsToMove)
-    looplike.moveOutOfLoop(op);
-  LLVM_DEBUG(looplike.print(llvm::dbgs() << "\n\nModified loop:\n"));
+
+  SmallVector<Value> values;
+  OperandRange lb_ops = loop.getLowerBoundsOperands(),
+               ub_ops = loop.getUpperBoundsOperands();
+  for (unsigned idx = 0; idx < loop.getLowerBoundsMap().getNumDims(); ++idx)
+    values.push_back(lb_ops[idx]);
+  for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumDims(); ++idx)
+    values.push_back(ub_ops[idx]);
+  for (unsigned idx = 0; idx < loop.getLowerBoundsMap().getNumSymbols(); ++idx)
+    values.push_back(lb_ops[idx + loop.getLowerBoundsMap().getNumDims()]);
+  for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumSymbols(); ++idx)
+    values.push_back(ub_ops[idx + loop.getUpperBoundsMap().getNumDims()]);
+
+  auto iset = IntegerSet::get(
+      /*dim*/ loop.getLowerBoundsMap().getNumDims() +
+          loop.getUpperBoundsMap().getNumDims(),
+      /*symbols*/ loop.getLowerBoundsMap().getNumSymbols() +
+          loop.getUpperBoundsMap().getNumSymbols(),
+      exprs, eqflags);
+  auto ifOp = b.create<AffineIfOp>(loop.getLoc(), TypeRange(), iset, values,
+                                   /*else*/ false);
+  loop->moveBefore(ifOp.getThenBlock()->getTerminator());
+
+  for (Operation *op : opsToMove)
+    loop.moveOutOfLoop(op);
+
+  LLVM_DEBUG(loop.print(llvm::dbgs() << "\n\nModified loop:\n"));
 }
 
 void ParallelLICM::runOnOperation() {
-  getOperation()->walk([&](LoopLikeOpInterface loopLike) {
-    LLVM_DEBUG(loopLike.print(llvm::dbgs() << "\nOriginal loop:\n"));
-    moveLoopInvariantCode(loopLike);
-    if (auto par = dyn_cast<scf::ParallelOp>((Operation *)loopLike)) {
-      moveParallelLoopInvariantCode(par);
-    } else if (auto par = dyn_cast<AffineParallelOp>((Operation *)loopLike)) {
-      moveParallelLoopInvariantCode(par);
-    }
+  getOperation()->walk([&](LoopLikeOpInterface loop) {
+    LLVM_DEBUG({
+      llvm::dbgs() << "----------------\n";
+      loop.print(llvm::dbgs() << "Original loop:\n");
+      llvm::dbgs() << "\n";
+    });
+
+    if (auto parLoop = dyn_cast<scf::ParallelOp>((Operation *)loop))
+      moveParallelLoopInvariantCode(parLoop);
+    else if (auto parLoop = dyn_cast<AffineParallelOp>((Operation *)loop))
+      moveParallelLoopInvariantCode(parLoop);
+    else
+      moveLoopInvariantCode(loop);
```

Comment:
This original code runs `moveLoopInvariantCode` to hoist all simple invariant code out of the loop first, why do you want to change that?

Reply:
Likely is not useful now given that `moveLoopInvariantCode` is more conservative, but `moveLoopInvariantCode` is from MLIR upstream, which can be improved, and could potentially hoist more general instructions then `moveParallelLoopInvariantCode`, and `moveParallelLoopInvariantCode` can focus on instructions that are special for parallel.

Reply:
Correct, is not useful currently. Things might change if the MLIR invariant code motion pass becomes more permissive, which would require loops to be guarded, which is not the case currently. 
I can put it back.